README
This is Business::OnlinePayment::CyberSource, a Business::OnlinePayment
backend module for CyberSource.  Information about CyberSource is available at
their website: http://www.cybersource.com

This module currently implements the 1.0.0 version of the Simple Order API 

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

To successfully run the tests, you will need to have the CyberSource Simple
Order API installed, and a working cybs.ini installed in the /etc directory.

This module requires these other modules and libraries:

  Simple Order API Perl (cybs)
  http://apps.cybersource.com/cgi-bin/pages/get_dir_dev_api.cgi?kit=Simple_Order_API_Perl

COPYRIGHT AND LICENCE

Copyright (C) 2004 by Peter Bowen

Written from nearly scratch, but based loosely on the Authorize.net module.

Business::OnlinePayment (required) is a generic interface for processing payments through
online credit card processors, online check acceptance houses, etc.  (If you
like buzzwords, call it an "multiplatform ecommerce-enabling middleware
solution").

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.3 or,
at your option, any later version of Perl 5 you may have available.







Changes
Revision history for {{$dist->name}}

{{$NEXT}}

3.000016  2014-03-05
	- finish converting to Module::Runtime
	- require Crypt::SSLeay for tests

3.000015  2014-03-04
	- add Business::OnlinePayment 3 as a Prereq

3.000014  2014-03-04
	- add conditional check to see if auth object exists

3.000013  2014-03-04
	- remove use of deprecated run_transaction API

3.000012  2014-03-03
	- convert to Module::Runtime
	- use Business::CyberSource 9 API
	- remove Data::Dump
	- update compile-tests should fix stall on windows
	- update metainformation

3.000011  2012-11-15
	- Change minimum version of B::C to 0.007006
	- Update MetaResources

3.000010  2012-10-29
	- Removed checking of accept role
	- removed debugging output

3.000009  2012-10-15
	- Remove unneeded git remote

3.000008  2012-10-15
	- Updated logic in Client class to rethrow exceptions thrown by the CyberSource response object

3.000007  2012-10-04
	- improve conditional checks for avs, authorization and cvcode

3.000006  2012-10-04
	- Add conditionals to only set avs_code and authorization if the appropriate roles are applied to the response object
	- Fix some logic errors in client object
	- Clean up some logic to remove unnecessary variables

3.000005  2012-09-04
	- Specify minimum version of Business::CyberSource as 0.006000
	- Set response values for success and failure
	- Add auth failure tests

3.000004  2012-08-30
	- fix dependencies

3.000003  2012-08-28
	- fix licensing
	- release to cpan

3.000002  2012-08-24
	- Fixed logic error preventing follow-on credits from working properly
	- Fixed Logic errors with expiration dates
	- Added support for expiration dates with MMYY format
	- various other bug fixes

3.000000  2012-08-16
	- Refactored code to use Business::CyberSource as the underlying engine

2.01      2011-07-25
	- use dist-zilla
	- mostly bring up to perl-critic compliance
	- allow passing a config hash on object initialization
	- fix to work with CyberSource Simple API 5.0
	- new maintainership Caleb Cushing

2.00	2007-04-21
	- release of 2.00, no changes noted, and I'm way to lazy to look at diff's

0.09	2007-03-01
	- Updated by Jason Hall (Jayce^) <jayce+cpan@lug-nut.com>
	- Updated to New Cybersource API
	- Specifically using the new 'requestToken' field that is necessary
	- Internal BOP option of 'security_key' to normalize its usage

0.08   2005-12-31
	- date unrecorded, so I made one up : Caleb Cushing
	- This update by Ivan Kohler <ivan-cybersource@420.am>
	- Add my previous patch against 0.6 to fix an explicitly specified
	  conf_file
	- Add "failure statuses" with a B:OP pre-3.00_04 workaround

0.07	2005-12-12
	- Improved the Documentation

0.06	2005-10-12
	- Allow both line items and grand total.  If you use both, the grand total will prevail.

0.05	2005-10-02
	- Set the defaults for the business logic settings using the cybs.ini

0.04	2005-08-09
	- Found a bug that allowed the system to point to production even if the cybs.conf said to use dev

0.03	2005-06-16
	- Ivan tested this and it was pretty broken.
	- Fixed the regexp for expiration date so that it works (Ivan patch)
		- Modified it slightly to allow for dates w/o slashes
	- Allowed for custom cybs.ini locations (Ivan Patch)
		- Made the default /etc/ as opposed to ./
	- Added a ton of error handling after Ivan had cats over the poor error
	  handling before
	  (Peter is being generous.  I was downright bitchy. -Ivan)

0.02	2005-01-31
	- First Release

0.01	2005-01-26
	- original version; created by hand.


lib/Business/OnlinePayment/CyberSource.pm
package Business::OnlinePayment::CyberSource;

use 5.010;
use strict;
use warnings;
use namespace::autoclean;

use Moose;
use Exception::Base;
use MooseX::NonMoose;
use MooseX::StrictConstructor;
use MooseX::Types::Moose qw(Bool HashRef Int);
use MooseX::Types::Common::String qw(NonEmptySimpleStr);

# ABSTRACT:  CyberSource backend for Business::OnlinePayment
# VERSION

extends 'Business::OnlinePayment';

#### Subroutine Definitions ####

# Post-construction hook
# Accepts:  A reference to a hash of construction parameters
# Returns:  Nothing

sub BUILD {
	my ( $self ) = @_;
	my $fields   = [ qw(type action reference_code amount) ];

	$self->required_fields( @$fields );

	return;
}

#### Object Attributes ####

#### Applied Roles ####

with 'Business::OnlinePayment::CyberSource::Role::TransactionHandling';

#### Method Modifiers ####

#### Meta class stuff ####

__PACKAGE__->meta->make_immutable();

1;

=pod

=head1 SYNOPSIS

	use Business::OnlinePayment;

	my $tx = Business::OnlinePayment->new( "CyberSource" );
	$tx->content(
		login          => 'username',
		password       => 'password',
		type           => 'CC',
		action         => 'Normal Authorization',
		invoice_number => '00000001', # MurchantReferenceCode
		first_name     => 'Peter',
		last_name      => 'Bowen',
		address        => '123 Anystreet',
		city           => 'Orem',
		state          => 'Utah',
		zip            => '84097',
		country        => 'US',
		email          => 'foo@bar.net',
		card_number    => '4111111111111111',
		expiration     => '09/06',
		cvv2           => '1234', #optional
		amount         => '5.00',
		currency       => 'USD',
	);

	$tx->submit();

	if($tx->is_success()) {
		print "Card processed successfully: ".$tx->authorization."\n";
	} else {
		print "Card was rejected: ".$tx->error_message."\n";
	}

	####
	# Two step transaction, authorization and capture.
	# If you don't need to review order before capture, you can
	# process in one step as above.
	####

  $tx = Business::OnlinePayment->new("CyberSource");
	$tx->content(
		login          => 'username',
		password       => 'password',
		type           => 'CC',
		action         => 'Authorization Only',
		invoice_number  => 44544, # MurchantReferenceCode
		description     => 'Business::OnlinePayment visa test',
		amount          => '42.39',
		first_name      => 'Tofu',
		last_name       => 'Beast',
		address         => '123 Anystreet',
		city            => 'Anywhere',
		state           => 'Utah',
		zip             => '84058',
		country         => 'US',
		email           => 'tofu@beast.org',
		card_number     => '4111111111111111',
		expiration      => '12/25',
		cvv2            => 1111,
	);
	$tx->submit();

	if($tx->is_success()) {
		# get information about authorization
		my $authorization = $tx->authorization();
		my $order_number = $tx->order_number(); # RequestId
		my $avs_code = $tx->avs_code(); # AVS Response Code();
		my $cvv2_response = $tx->cvv2_response(); # CVV2/CVC2/CID Response Code();

		# now capture transaction

		$tx->content(
			login          => 'username',
			password       => 'password',
			type           => 'CC',
			action         => 'Post Authorization',
			invoice_number => 44544, #MurchantReferenceCode
			amount         => '42.39',
			po_number       => $tx->order_number(), # RequestId
		);

		$tx->submit();

		if($tx->is_success()) {
			print "Funds captured successfully\n";
		} else {
			print "Card was rejected: ".$tx->error_message."\n";
		}

	} else {
		print "Card was rejected: " . $tx->error_message() . "\n";
	}

=head1 DESCRIPTION

For detailed information see L<Business::OnlinePayment>.

=head1 SUPPORTED TRANSACTION TYPES

=head2 CC

Content required: type, login, action, amount, first_name, last_name, card_number, expiration.

cvv2 is required in order to get back a cvv2_response value.

=head2 Settling

To settle an authorization-only transaction (where you set action to
C<Authorization Only>), submit the C<order_number> code in the field
C<po_number> with the action set to C<Post Authorization>.

You can get the transaction id from the authorization by calling the
C<order_number> method on the object returned from the authorization.
You must also submit the amount field with a value less than or equal
to the amount specified in the original authorization.

=method BUILD

this is a before-construction hook for Moose.  You Will never call this method directly.

=head1 ACKNOWLEDGMENTS

=over 4

=item Jason Kohles

For writing BOP - I didn't have to create my own framework.

=item Ivan Kohler

Tested the first pre-release version and fixed a number of bugs.
He also encouraged me to add better error reporting for system
errors.  He also added failure_status support.

=item Jason (Jayce^) Hall

Adding Request Token Requirements (Among other significant improvements... )

=back

=head1 SEE ALSO

L<Business::OnlinePayment>

=cut
